File: data-structures-in-c/tree/bst.c
#include <stdio.h>
#include <stdlib.h>

typedef struct Node {
  int data;
  struct Node *left;
  struct Node *right;
} Node;

Node *root = NULL;
int count = 0;

void insert(int data, Node *root) {
  if (!root) {
    root = (Node *)malloc(sizeof(Node));
    root->data = data;
    root->left = NULL;
    root->right = NULL;
    count++;
    return;
  }

  if (root->data < data) {
    if (root->left) {
      insert(data, root->left);
      return;
    }

    root->left = (Node *)malloc(sizeof(Node));
    root->left->data = data;
    root->left->left = NULL;
    root->left->right = NULL;
    count++;
    return;
  }

  if (root->right) {
    insert(data, root->right);
    return;
  }

  root->right = (Node *)malloc(sizeof(Node));
  root->right->data = data;
  root->right->left = NULL;
  root->right->right = NULL;
  count++;
}

int main() {
  insert(10, root);
  insert(20, root);
  insert(30, root);
  insert(40, root);

  printf("total elements: %d\n", count);

  return 0;
}

File: go/data-structures/bst/bst.go
package bst

import (
	"fmt"
	"strings"

	"golang.org/x/exp/constraints"
)

type Tree[T constraints.Ordered] struct {
	root *Node[T]
}

type Node[T constraints.Ordered] struct {
	Datum T
	Left  *Node[T]
	Right *Node[T]
}

func New[T constraints.Ordered]() Tree[T] {
	return Tree[T]{root: nil}
}

func (b Tree[T]) String() string {
	var str string
	b.traverse(&str, b.root)
	str = strings.TrimSpace(str)
	if str == "" {
		return "nil"
	}
	return str
}

func (b *Tree[T]) Insert(data ...T) {
	for _, datum := range data {
		b.insertOne(datum, b.root)
	}
}

func (b Tree[T]) Search(datum T) bool {
	if b.root == nil {
		return false
	}
	return b.root.search(datum)
}

func (n Node[T]) search(datum T) bool {
	if datum < n.Datum {
		if n.Left == nil {
			return false
		}
		return n.Left.search(datum)
	}
	if datum > n.Datum {
		if n.Right == nil {
			return false
		}
		return n.Right.search(datum)
	}
	return true
}

func (b Tree[T]) traverse(str *string, n *Node[T]) {
	if n == nil {
		return
	}
	*str += fmt.Sprintf("%v -> %s\n", n.Datum, n.asSet())
	b.traverse(str, n.Left)
	b.traverse(str, n.Right)
}

func (n Node[T]) asSet() string {
	left := "nil"
	if n.Left != nil {
		left = fmt.Sprint(n.Left.Datum)
	}
	right := "nil"
	if n.Right != nil {
		right = fmt.Sprint(n.Right.Datum)
	}
	return fmt.Sprintf("(%s, %s)", left, right)
}

func (b *Tree[T]) insertOne(datum T, n *Node[T]) {
	if b.root == nil {
		b.root = &Node[T]{Datum: datum}
		return
	}

	if datum < n.Datum {
		if n.Left == nil {
			n.Left = &Node[T]{Datum: datum}
			return
		}
		b.insertOne(datum, n.Left)
		return
	}

	if n.Right == nil {
		n.Right = &Node[T]{Datum: datum}
		return
	}
	b.insertOne(datum, n.Right)
}
